<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9 Heatmaps | R Tutorials</title>
  <meta name="description" content="R tutorials for beginners!" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="9 Heatmaps | R Tutorials" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="/cover.png" />
  <meta property="og:description" content="R tutorials for beginners!" />
  <meta name="github-repo" content="kazeera/r-seesions" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9 Heatmaps | R Tutorials" />
  
  <meta name="twitter:description" content="R tutorials for beginners!" />
  <meta name="twitter:image" content="/cover.png" />

<meta name="author" content="Kazeera Aliar" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="correlation-analysis.html"/>
<link rel="next" href="dataframe-manipulation.html"/>
<script src="libs/header-attrs-2.8/header-attrs.js"></script>
<script src="libs/jquery-3.5.1/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script>
<script src="libs/plotly-binding-4.9.4.1/plotly.js"></script>
<script src="libs/typedarray-0.1/typedarray.min.js"></script>
<link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet" />
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script>
<link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="install-r.html"><a href="install-r.html"><i class="fa fa-check"></i><b>2</b> Install R</a>
<ul>
<li class="chapter" data-level="2.1" data-path="install-r.html"><a href="install-r.html#download-r"><i class="fa fa-check"></i><b>2.1</b> Download R</a></li>
<li class="chapter" data-level="2.2" data-path="install-r.html"><a href="install-r.html#download-rstudio"><i class="fa fa-check"></i><b>2.2</b> Download RStudio</a></li>
<li class="chapter" data-level="2.3" data-path="install-r.html"><a href="install-r.html#terms"><i class="fa fa-check"></i><b>2.3</b> Terms</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="r-programming-basics.html"><a href="r-programming-basics.html"><i class="fa fa-check"></i><b>3</b> R Programming Basics</a>
<ul>
<li class="chapter" data-level="3.1" data-path="r-programming-basics.html"><a href="r-programming-basics.html#variable-declaration"><i class="fa fa-check"></i><b>3.1</b> Variable Declaration</a></li>
<li class="chapter" data-level="3.2" data-path="r-programming-basics.html"><a href="r-programming-basics.html#basic-data-types"><i class="fa fa-check"></i><b>3.2</b> Basic Data Types</a></li>
<li class="chapter" data-level="3.3" data-path="r-programming-basics.html"><a href="r-programming-basics.html#operators"><i class="fa fa-check"></i><b>3.3</b> Operators</a></li>
<li class="chapter" data-level="3.4" data-path="r-programming-basics.html"><a href="r-programming-basics.html#practice"><i class="fa fa-check"></i><b>3.4</b> Practice</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>4</b> Data Structures</a>
<ul>
<li class="chapter" data-level="4.1" data-path="data-structures.html"><a href="data-structures.html#vectors"><i class="fa fa-check"></i><b>4.1</b> Vectors</a></li>
<li class="chapter" data-level="4.2" data-path="data-structures.html"><a href="data-structures.html#data-frames"><i class="fa fa-check"></i><b>4.2</b> Data Frames</a></li>
<li class="chapter" data-level="4.3" data-path="data-structures.html"><a href="data-structures.html#practice-1"><i class="fa fa-check"></i><b>4.3</b> Practice</a></li>
<li class="chapter" data-level="4.4" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>4.4</b> Lists</a></li>
<li class="chapter" data-level="4.5" data-path="data-structures.html"><a href="data-structures.html#factors"><i class="fa fa-check"></i><b>4.5</b> Factors</a></li>
<li class="chapter" data-level="4.6" data-path="data-structures.html"><a href="data-structures.html#matrices"><i class="fa fa-check"></i><b>4.6</b> Matrices</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="importexport.html"><a href="importexport.html"><i class="fa fa-check"></i><b>5</b> Import/Export</a>
<ul>
<li class="chapter" data-level="5.1" data-path="importexport.html"><a href="importexport.html#export-tables"><i class="fa fa-check"></i><b>5.1</b> Export Tables</a></li>
<li class="chapter" data-level="5.2" data-path="importexport.html"><a href="importexport.html#import-tables"><i class="fa fa-check"></i><b>5.2</b> Import Tables</a></li>
<li class="chapter" data-level="5.3" data-path="importexport.html"><a href="importexport.html#graphing-with-base-r"><i class="fa fa-check"></i><b>5.3</b> Graphing with base R</a></li>
<li class="chapter" data-level="5.4" data-path="importexport.html"><a href="importexport.html#practice-2"><i class="fa fa-check"></i><b>5.4</b> Practice</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="graphing-with-ggplot2.html"><a href="graphing-with-ggplot2.html"><i class="fa fa-check"></i><b>6</b> Graphing with ggplot2</a>
<ul>
<li class="chapter" data-level="6.1" data-path="graphing-with-ggplot2.html"><a href="graphing-with-ggplot2.html#what-is-ggplot2"><i class="fa fa-check"></i><b>6.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="6.2" data-path="graphing-with-ggplot2.html"><a href="graphing-with-ggplot2.html#factors-1"><i class="fa fa-check"></i><b>6.2</b> Factors</a></li>
<li class="chapter" data-level="6.3" data-path="graphing-with-ggplot2.html"><a href="graphing-with-ggplot2.html#r-colors"><i class="fa fa-check"></i><b>6.3</b> R Colors</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="statistical-tests.html"><a href="statistical-tests.html"><i class="fa fa-check"></i><b>7</b> Statistical Tests</a>
<ul>
<li class="chapter" data-level="7.1" data-path="statistical-tests.html"><a href="statistical-tests.html#lists-1"><i class="fa fa-check"></i><b>7.1</b> Lists</a></li>
<li class="chapter" data-level="7.2" data-path="statistical-tests.html"><a href="statistical-tests.html#statistical-significance"><i class="fa fa-check"></i><b>7.2</b> Statistical Significance</a></li>
<li class="chapter" data-level="7.3" data-path="statistical-tests.html"><a href="statistical-tests.html#checks-for-normality"><i class="fa fa-check"></i><b>7.3</b> Checks for Normality</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="statistical-tests.html"><a href="statistical-tests.html#shapiro-test"><i class="fa fa-check"></i><b>7.3.1</b> Shapiro Test</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="statistical-tests.html"><a href="statistical-tests.html#one-sample-tests"><i class="fa fa-check"></i><b>7.4</b> One-Sample Tests</a></li>
<li class="chapter" data-level="7.5" data-path="statistical-tests.html"><a href="statistical-tests.html#two-sample-tests"><i class="fa fa-check"></i><b>7.5</b> Two-Sample Tests</a></li>
<li class="chapter" data-level="7.6" data-path="statistical-tests.html"><a href="statistical-tests.html#anova"><i class="fa fa-check"></i><b>7.6</b> ANOVA</a></li>
<li class="chapter" data-level="7.7" data-path="statistical-tests.html"><a href="statistical-tests.html#adding-p-values-to-ggplot"><i class="fa fa-check"></i><b>7.7</b> Adding p-values to ggplot</a></li>
<li class="chapter" data-level="7.8" data-path="statistical-tests.html"><a href="statistical-tests.html#practice-3"><i class="fa fa-check"></i><b>7.8</b> Practice</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="correlation-analysis.html"><a href="correlation-analysis.html"><i class="fa fa-check"></i><b>8</b> Correlation Analysis</a>
<ul>
<li class="chapter" data-level="8.1" data-path="correlation-analysis.html"><a href="correlation-analysis.html#correlation-methods"><i class="fa fa-check"></i><b>8.1</b> Correlation Methods</a></li>
<li class="chapter" data-level="8.2" data-path="correlation-analysis.html"><a href="correlation-analysis.html#correlation-matrix"><i class="fa fa-check"></i><b>8.2</b> Correlation Matrix:</a></li>
<li class="chapter" data-level="8.3" data-path="correlation-analysis.html"><a href="correlation-analysis.html#practice-correlation-analysis"><i class="fa fa-check"></i><b>8.3</b> Practice: Correlation analysis</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="heatmaps.html"><a href="heatmaps.html"><i class="fa fa-check"></i><b>9</b> Heatmaps</a>
<ul>
<li class="chapter" data-level="9.1" data-path="heatmaps.html"><a href="heatmaps.html#data-preparation"><i class="fa fa-check"></i><b>9.1</b> Data preparation</a></li>
<li class="chapter" data-level="9.2" data-path="heatmaps.html"><a href="heatmaps.html#pheatmap"><i class="fa fa-check"></i><b>9.2</b> pheatmap()</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="heatmaps.html"><a href="heatmaps.html#modify-row-and-column-names"><i class="fa fa-check"></i><b>9.2.1</b> Modify row and column names</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="heatmaps.html"><a href="heatmaps.html#scaling"><i class="fa fa-check"></i><b>9.3</b> Scaling</a></li>
<li class="chapter" data-level="9.4" data-path="heatmaps.html"><a href="heatmaps.html#heatmap-colours"><i class="fa fa-check"></i><b>9.4</b> Heatmap Colours</a></li>
<li class="chapter" data-level="9.5" data-path="heatmaps.html"><a href="heatmaps.html#clustering-methods"><i class="fa fa-check"></i><b>9.5</b> Clustering methods</a></li>
<li class="chapter" data-level="9.6" data-path="heatmaps.html"><a href="heatmaps.html#create-annotations"><i class="fa fa-check"></i><b>9.6</b> Create annotations</a></li>
<li class="chapter" data-level="9.7" data-path="heatmaps.html"><a href="heatmaps.html#practice-4"><i class="fa fa-check"></i><b>9.7</b> Practice</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="dataframe-manipulation.html"><a href="dataframe-manipulation.html"><i class="fa fa-check"></i><b>10</b> Dataframe Manipulation</a>
<ul>
<li class="chapter" data-level="10.1" data-path="dataframe-manipulation.html"><a href="dataframe-manipulation.html#reshape-data"><i class="fa fa-check"></i><b>10.1</b> Reshape Data</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="dataframe-manipulation.html"><a href="dataframe-manipulation.html#data"><i class="fa fa-check"></i><b>10.1.1</b> Data</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="dataframe-manipulation.html"><a href="dataframe-manipulation.html#pipe-operator"><i class="fa fa-check"></i><b>10.2</b> Pipe operator</a></li>
<li class="chapter" data-level="10.3" data-path="dataframe-manipulation.html"><a href="dataframe-manipulation.html#practice-5"><i class="fa fa-check"></i><b>10.3</b> Practice</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="search-and-conditions.html"><a href="search-and-conditions.html"><i class="fa fa-check"></i><b>11</b> Search and Conditions</a>
<ul>
<li class="chapter" data-level="11.1" data-path="search-and-conditions.html"><a href="search-and-conditions.html#general-expressions"><i class="fa fa-check"></i><b>11.1</b> General Expressions</a></li>
<li class="chapter" data-level="11.2" data-path="search-and-conditions.html"><a href="search-and-conditions.html#conditions"><i class="fa fa-check"></i><b>11.2</b> Conditions</a></li>
<li class="chapter" data-level="11.3" data-path="search-and-conditions.html"><a href="search-and-conditions.html#any-and-all"><i class="fa fa-check"></i><b>11.3</b> any() and all()</a></li>
<li class="chapter" data-level="11.4" data-path="search-and-conditions.html"><a href="search-and-conditions.html#operators-1"><i class="fa fa-check"></i><b>11.4</b> Operators</a></li>
<li class="chapter" data-level="11.5" data-path="search-and-conditions.html"><a href="search-and-conditions.html#ifelse"><i class="fa fa-check"></i><b>11.5</b> ifelse()</a></li>
<li class="chapter" data-level="11.6" data-path="search-and-conditions.html"><a href="search-and-conditions.html#practice-6"><i class="fa fa-check"></i><b>11.6</b> Practice</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="repetitive-jobs.html"><a href="repetitive-jobs.html"><i class="fa fa-check"></i><b>12</b> Repetitive Jobs</a>
<ul>
<li class="chapter" data-level="12.1" data-path="repetitive-jobs.html"><a href="repetitive-jobs.html#functions"><i class="fa fa-check"></i><b>12.1</b> Functions</a></li>
<li class="chapter" data-level="12.2" data-path="repetitive-jobs.html"><a href="repetitive-jobs.html#loops"><i class="fa fa-check"></i><b>12.2</b> Loops</a></li>
<li class="chapter" data-level="12.3" data-path="repetitive-jobs.html"><a href="repetitive-jobs.html#vectorization"><i class="fa fa-check"></i><b>12.3</b> Vectorization</a></li>
<li class="chapter" data-level="12.4" data-path="repetitive-jobs.html"><a href="repetitive-jobs.html#intro-to-apply"><i class="fa fa-check"></i><b>12.4</b> Intro to apply</a></li>
<li class="chapter" data-level="12.5" data-path="repetitive-jobs.html"><a href="repetitive-jobs.html#practice-7"><i class="fa fa-check"></i><b>12.5</b> Practice</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R Tutorials</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="heatmaps" class="section level1" number="9">
<h1><span class="header-section-number">9</span> Heatmaps</h1>
<p>Objective: To learn how to draw heatmaps in R using the pheatmap package</p>
<p>Summary:
- Data preparation:
- Numeric matrix/data frame as input
- Log normalization
- Making heatmaps (base R heatmap() function and pheatmap())
- Customization using arguments
- Scaling
- Clustering
- Adding annotations (columns and rows)</p>
<p>This session is based upon this tutorial: datanovia.com/en/lessons/heatmap-in-r-static-and-interactive-visualization/</p>
<p>What is a heatmap?
- plot to simultaneously visualize clusters of samples and features (i.e. hierarchical clustering)
- blocks of high and low values are adjacent
- “false colored image”, where data values are transformed to color scale</p>
<p>R Packages/functions for drawing heatmaps
* heatmap() [R base function, stats package]: Draws a simple heatmap
* heatmap.2() [gplots R package]: Draws an enhanced heatmap compared to the R base function.
* pheatmap() [pheatmap R package]: Draws pretty heatmaps and provides more control to change the appearance of heatmaps.
* d3heatmap() [d3heatmap R package]: Draws an interactive/clickable heatmap
* Heatmap() [ComplexHeatmap R/Bioconductor package]: Draws, annotates and arranges complex heatmaps (very useful for genomic data analysis)</p>
<div id="data-preparation" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> Data preparation</h2>
<p>The proteomic data we’re using today is from Alison Casey’s 2018 JCB paper. We will look at the expression of 304 proteins.</p>
<p>Note: the column names are in the format “Celltype_HormoneTreatment_Replicate” where
for celltype, BC = basal cell,
LP = luminal progenitor,
LM = luminal mature
for hormone treatment,
E = estrogen,
EP = estrogen-progesterone</p>
<p>Read the tab-delimited text file.</p>
<div class="sourceCode" id="cb428"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb428-1"><a href="heatmaps.html#cb428-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read into data frame object using read.delim()</span></span>
<span id="cb428-2"><a href="heatmaps.html#cb428-2" aria-hidden="true" tabindex="-1"></a>exp_df <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="at">file =</span> <span class="st">&quot;07-Casey-mamm-mouse-proteome-sample.txt&quot;</span>, <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb428-3"><a href="heatmaps.html#cb428-3" aria-hidden="true" tabindex="-1"></a><span class="co"># See the dimensions (row x col)</span></span>
<span id="cb428-4"><a href="heatmaps.html#cb428-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(exp_df)</span></code></pre></div>
<pre><code>## [1] 304  13</code></pre>
<div class="sourceCode" id="cb430"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb430-1"><a href="heatmaps.html#cb430-1" aria-hidden="true" tabindex="-1"></a><span class="co"># See the first 6 rows</span></span>
<span id="cb430-2"><a href="heatmaps.html#cb430-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(exp_df)</span></code></pre></div>
<pre><code>##   mGeneSym   BC_E_1  BC_E_2  BC_EP_1  BC_EP_2   LP_E_1   LP_E_2  LP_EP_1
## 1   Abcb10        0       0        0        0  1557708  1013752  4221663
## 2    Abcg2        0       0  2163991  1048502 12987493  7697484  6474732
## 3    Abhd4  4956153 5124382 13346955 10634977        0        0        0
## 4    Abhd5        0       0        0        0  1255272  4329562  5741534
## 5    Acox1  8926818 9756252  7242814  6847158  4069287  6442152 19309447
## 6      Ada 12920739 9339090  3126248  8926815 45120667 77396731 80759779
##    LP_EP_2   LM_E_1     LM_E_2   LM_EP_1   LM_EP_2
## 1  5882079  1235621   802389.8   2127015   3010426
## 2 13136088        0  5888476.7         0         0
## 3        0  5872301  8928660.8   2478597   6995790
## 4  2653241 14846810 39848291.1   8631368   6878478
## 5 11300060  5050705  6733985.2         0         0
## 6 87393916 65529106 69778380.1 162933067 211765680</code></pre>
<p><strong>Numeric Matrices</strong></p>
<p>Most heat map function require that the data be in the form of a numeric matrix, where the rows and columns are genes and samples.<br />
Recall: matrices are data tables, where all columns and rows have the same data type.</p>
<p>Note: data frames are ok for pheatmap(), as long as all the columns are numeric (right now, our data is not all numeric).</p>
<div class="sourceCode" id="cb432"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb432-1"><a href="heatmaps.html#cb432-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if an object is a matrix</span></span>
<span id="cb432-2"><a href="heatmaps.html#cb432-2" aria-hidden="true" tabindex="-1"></a><span class="fu">is.matrix</span>(exp_df)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb434"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb434-1"><a href="heatmaps.html#cb434-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename rows with feature (ie. gene name) and remove the feature column</span></span>
<span id="cb434-2"><a href="heatmaps.html#cb434-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(exp_df) <span class="ot">&lt;-</span> exp_df<span class="sc">$</span>mGeneSym</span>
<span id="cb434-3"><a href="heatmaps.html#cb434-3" aria-hidden="true" tabindex="-1"></a>exp_df<span class="sc">$</span>mGeneSym <span class="ot">&lt;-</span> <span class="cn">NULL</span> <span class="co"># equivalent code: exp_df &lt;- exp_df[,-1]</span></span>
<span id="cb434-4"><a href="heatmaps.html#cb434-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb434-5"><a href="heatmaps.html#cb434-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert a data frame to a data matrix</span></span>
<span id="cb434-6"><a href="heatmaps.html#cb434-6" aria-hidden="true" tabindex="-1"></a>exp_mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(exp_df)</span>
<span id="cb434-7"><a href="heatmaps.html#cb434-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(exp_mat)</span></code></pre></div>
<pre><code>##          BC_E_1  BC_E_2  BC_EP_1  BC_EP_2   LP_E_1   LP_E_2  LP_EP_1  LP_EP_2
## Abcb10        0       0        0        0  1557708  1013752  4221663  5882079
## Abcg2         0       0  2163991  1048502 12987493  7697484  6474732 13136088
## Abhd4   4956153 5124382 13346955 10634977        0        0        0        0
## Abhd5         0       0        0        0  1255272  4329562  5741534  2653241
## Acox1   8926818 9756252  7242814  6847158  4069287  6442152 19309447 11300060
## Ada    12920739 9339090  3126248  8926815 45120667 77396731 80759779 87393916
##          LM_E_1     LM_E_2   LM_EP_1   LM_EP_2
## Abcb10  1235621   802389.8   2127015   3010426
## Abcg2         0  5888476.7         0         0
## Abhd4   5872301  8928660.8   2478597   6995790
## Abhd5  14846810 39848291.1   8631368   6878478
## Acox1   5050705  6733985.2         0         0
## Ada    65529106 69778380.1 162933067 211765680</code></pre>
<p><strong>Log normalization</strong></p>
<p>Usually, expression data has large ranges of values. This will be scaled in the heatmap but we may lose sight of small differences.
And so, we log our data.</p>
<p>We must deal with the zeros before log(0) is negative infinity (can’t plot -Inf).</p>
<div class="sourceCode" id="cb436"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb436-1"><a href="heatmaps.html#cb436-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if there are any zeros in the data</span></span>
<span id="cb436-2"><a href="heatmaps.html#cb436-2" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(exp_mat <span class="sc">==</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb438"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb438-1"><a href="heatmaps.html#cb438-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Impute 0 with 1</span></span>
<span id="cb438-2"><a href="heatmaps.html#cb438-2" aria-hidden="true" tabindex="-1"></a>exp_mat[exp_mat <span class="sc">==</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb438-3"><a href="heatmaps.html#cb438-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Log2-transform data </span></span>
<span id="cb438-4"><a href="heatmaps.html#cb438-4" aria-hidden="true" tabindex="-1"></a>exp_mat <span class="ot">&lt;-</span> <span class="fu">log2</span>(exp_mat)</span>
<span id="cb438-5"><a href="heatmaps.html#cb438-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb438-6"><a href="heatmaps.html#cb438-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Alternatively, log2(exp_mat + 1) adds 1 to all values before log transformation</span></span></code></pre></div>
<p>Now let’s make our heatmaps..</p>
<p><strong>R base heatmap: heatmap()</strong></p>
<p>The built-in R heatmap() function [in stats package] can be used.</p>
<div class="sourceCode" id="cb439"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb439-1"><a href="heatmaps.html#cb439-1" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(<span class="at">x =</span> exp_mat) <span class="co"># x is a numeric matrix, high values are in red and low values are in yellow</span></span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-86-1.png" width="672" /></p>
</div>
<div id="pheatmap" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> pheatmap()</h2>
<p>Use the pheatmap (“pretty” heat map package) for more customizations.</p>
<div class="sourceCode" id="cb440"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb440-1"><a href="heatmaps.html#cb440-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install and load package</span></span>
<span id="cb440-2"><a href="heatmaps.html#cb440-2" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;pheatmap&quot;)</span></span>
<span id="cb440-3"><a href="heatmaps.html#cb440-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb440-4"><a href="heatmaps.html#cb440-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb440-5"><a href="heatmaps.html#cb440-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Make a simple heatmap using pheatmap() function call</span></span>
<span id="cb440-6"><a href="heatmaps.html#cb440-6" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat) <span class="co"># input: numeric matrix</span></span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-87-1.png" width="672" /></p>
<div id="modify-row-and-column-names" class="section level3" number="9.2.1">
<h3><span class="header-section-number">9.2.1</span> Modify row and column names</h3>
<p>pheatmap allows you to specify font size, change the row/column names, remove the names, and much more.</p>
<div class="sourceCode" id="cb441"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb441-1"><a href="heatmaps.html#cb441-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, </span>
<span id="cb441-2"><a href="heatmaps.html#cb441-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_rownames =</span> F, <span class="co"># hide row names</span></span>
<span id="cb441-3"><a href="heatmaps.html#cb441-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontsize_col =</span> <span class="dv">2</span>, <span class="co"># fontsize of column labels (sample names)</span></span>
<span id="cb441-4"><a href="heatmaps.html#cb441-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">cellwidth =</span> <span class="dv">3</span>) <span class="co"># width of columns</span></span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-88-1.png" width="672" /></p>
</div>
</div>
<div id="scaling" class="section level2" number="9.3">
<h2><span class="header-section-number">9.3</span> Scaling</h2>
<p>Usually, we scale by our features (genes) to normalize gene expression across different samples.
- argument built in to heatmap function call
- a character indicating if the values should be centered and scaled in the “row” direction, “column” direction, or “none”</p>
<div class="sourceCode" id="cb442"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb442-1"><a href="heatmaps.html#cb442-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, <span class="at">show_rownames =</span> F, </span>
<span id="cb442-2"><a href="heatmaps.html#cb442-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>) <span class="co">#default scale</span></span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-89-1.png" width="672" /></p>
<p>You can scale the expression matrix before plotting the heatmap as well.</p>
<ul>
<li>compute z-scores - Z-score normalization is a strategy of normalizing data that avoids outlier issue by considering mean and standard deviation
Read more: <a href="https://www.codecademy.com/articles/normalization" class="uri">https://www.codecademy.com/articles/normalization</a></li>
</ul>
<div class="sourceCode" id="cb443"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb443-1"><a href="heatmaps.html#cb443-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute z-scores to plot on heat map</span></span>
<span id="cb443-2"><a href="heatmaps.html#cb443-2" aria-hidden="true" tabindex="-1"></a><span class="co"># apply() performs a computation/function over rows (1) or columns (2) in a dataframe or matrix</span></span>
<span id="cb443-3"><a href="heatmaps.html#cb443-3" aria-hidden="true" tabindex="-1"></a>z_exp_mat <span class="ot">&lt;-</span> <span class="fu">apply</span>(exp_mat, <span class="dv">1</span>, <span class="cf">function</span>(x) (x <span class="sc">-</span> <span class="fu">mean</span>(x)) <span class="sc">/</span> <span class="fu">sd</span>(x)) </span>
<span id="cb443-4"><a href="heatmaps.html#cb443-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb443-5"><a href="heatmaps.html#cb443-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Transpose matrix, since apply() rotates data frame</span></span>
<span id="cb443-6"><a href="heatmaps.html#cb443-6" aria-hidden="true" tabindex="-1"></a>z_exp_mat <span class="ot">&lt;-</span> <span class="fu">t</span>(z_exp_mat)</span>
<span id="cb443-7"><a href="heatmaps.html#cb443-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb443-8"><a href="heatmaps.html#cb443-8" aria-hidden="true" tabindex="-1"></a><span class="co"># This is your input data matrix</span></span>
<span id="cb443-9"><a href="heatmaps.html#cb443-9" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(z_exp_mat, <span class="at">show_rownames =</span> F, <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>) </span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-90-1.png" width="672" /></p>
<ul>
<li>rescale to specified min/max using rescale() from the scales package
e.g. make range of {-3.2 to 4} to {-2 to 2})</li>
</ul>
<div class="sourceCode" id="cb444"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb444-1"><a href="heatmaps.html#cb444-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;scales&quot;)</span></span>
<span id="cb444-2"><a href="heatmaps.html#cb444-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb444-3"><a href="heatmaps.html#cb444-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at range (minimum and maximum)</span></span>
<span id="cb444-4"><a href="heatmaps.html#cb444-4" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(exp_mat)</span></code></pre></div>
<pre><code>## [1]  0.00000 34.53543</code></pre>
<div class="sourceCode" id="cb446"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb446-1"><a href="heatmaps.html#cb446-1" aria-hidden="true" tabindex="-1"></a><span class="co"># -- rescale whole matrix</span></span>
<span id="cb446-2"><a href="heatmaps.html#cb446-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Rescale whole matrix - this just looks like the first heatmap we made</span></span>
<span id="cb446-3"><a href="heatmaps.html#cb446-3" aria-hidden="true" tabindex="-1"></a>exp_mat2 <span class="ot">&lt;-</span> <span class="fu">rescale</span>(<span class="at">x =</span> exp_mat, <span class="co"># numeric matrix</span></span>
<span id="cb446-4"><a href="heatmaps.html#cb446-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">to =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">2</span>)) <span class="co"># new range</span></span>
<span id="cb446-5"><a href="heatmaps.html#cb446-5" aria-hidden="true" tabindex="-1"></a><span class="co"># In your heatmap function call, specify the scale = &quot;none&quot; to preserve the range</span></span>
<span id="cb446-6"><a href="heatmaps.html#cb446-6" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat2, <span class="at">show_rownames =</span> F, <span class="at">scale =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-91-1.png" width="672" /></p>
<div class="sourceCode" id="cb447"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb447-1"><a href="heatmaps.html#cb447-1" aria-hidden="true" tabindex="-1"></a><span class="co"># -- rescale each row seperately</span></span>
<span id="cb447-2"><a href="heatmaps.html#cb447-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Rescale each row seperately</span></span>
<span id="cb447-3"><a href="heatmaps.html#cb447-3" aria-hidden="true" tabindex="-1"></a>exp_mat2 <span class="ot">&lt;-</span> <span class="fu">apply</span>(exp_mat, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="fu">rescale</span>(x, <span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>, <span class="dv">2</span>)))</span>
<span id="cb447-4"><a href="heatmaps.html#cb447-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Transpose when using the apply function</span></span>
<span id="cb447-5"><a href="heatmaps.html#cb447-5" aria-hidden="true" tabindex="-1"></a>exp_mat2 <span class="ot">&lt;-</span> <span class="fu">t</span>(exp_mat2)</span>
<span id="cb447-6"><a href="heatmaps.html#cb447-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot heatmap - remember to transpose when </span></span>
<span id="cb447-7"><a href="heatmaps.html#cb447-7" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat2, <span class="at">show_rownames =</span> F, <span class="at">scale =</span> <span class="st">&quot;none&quot;</span>)</span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-91-2.png" width="672" /></p>
<div class="sourceCode" id="cb448"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb448-1"><a href="heatmaps.html#cb448-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Look at new range (minimum and maximum)</span></span>
<span id="cb448-2"><a href="heatmaps.html#cb448-2" aria-hidden="true" tabindex="-1"></a><span class="fu">range</span>(exp_mat2)</span></code></pre></div>
<pre><code>## [1] -2  2</code></pre>
</div>
<div id="heatmap-colours" class="section level2" number="9.4">
<h2><span class="header-section-number">9.4</span> Heatmap Colours</h2>
<ul>
<li>It’s possible to specify a color palette using the argument col, which can be defined as follow:</li>
<li>use colorRampPalette() to make a palette and “multiply” by a number to create a gradient</li>
</ul>
<div class="sourceCode" id="cb450"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb450-1"><a href="heatmaps.html#cb450-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Using custom colors:</span></span>
<span id="cb450-2"><a href="heatmaps.html#cb450-2" aria-hidden="true" tabindex="-1"></a>col_palette <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;blue&quot;</span>)) (<span class="dv">256</span>) <span class="co"># 256 is the number of colours this function will create</span></span>
<span id="cb450-3"><a href="heatmaps.html#cb450-3" aria-hidden="true" tabindex="-1"></a>scales<span class="sc">::</span><span class="fu">show_col</span>(col_palette) <span class="co">#show colour palette</span></span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-92-1.png" width="672" /></p>
<div class="sourceCode" id="cb451"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb451-1"><a href="heatmaps.html#cb451-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Using RColorBrewer color palette:</span></span>
<span id="cb451-2"><a href="heatmaps.html#cb451-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;RColorBrewer&quot;</span>)</span>
<span id="cb451-3"><a href="heatmaps.html#cb451-3" aria-hidden="true" tabindex="-1"></a>RColorBrewer<span class="sc">::</span><span class="fu">display.brewer.all</span>()</span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-92-2.png" width="672" /></p>
<div class="sourceCode" id="cb452"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb452-1"><a href="heatmaps.html#cb452-1" aria-hidden="true" tabindex="-1"></a>col_palette <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>, <span class="st">&quot;PuOr&quot;</span>)) (<span class="dv">256</span>) <span class="co">#pick a palette and multiply by 256 using ()</span></span>
<span id="cb452-2"><a href="heatmaps.html#cb452-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb452-3"><a href="heatmaps.html#cb452-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Show on heatmap</span></span>
<span id="cb452-4"><a href="heatmaps.html#cb452-4" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, <span class="at">show_rownames =</span> F, <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>, </span>
<span id="cb452-5"><a href="heatmaps.html#cb452-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">col =</span>  col_palette) <span class="co"># specify colour palette</span></span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-92-3.png" width="672" /></p>
</div>
<div id="clustering-methods" class="section level2" number="9.5">
<h2><span class="header-section-number">9.5</span> Clustering methods</h2>
<ul>
<li>clustering refers to visualizing groups of similar values together via a dendogram (tree-like object) and is always unsupervised in heatmaps</li>
<li>rows and columns are clustered independent of one another</li>
</ul>
<p>Set “cluster_cols” and “cluster_rows” to FALSE to prevent clustering. Default is TRUE (via euclidean distance and complete linkage).</p>
<div class="sourceCode" id="cb453"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb453-1"><a href="heatmaps.html#cb453-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, <span class="at">show_rownames =</span> F, <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>, <span class="at">col =</span>  col_palette, </span>
<span id="cb453-2"><a href="heatmaps.html#cb453-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols =</span> F, <span class="co"># specify column clustering</span></span>
<span id="cb453-3"><a href="heatmaps.html#cb453-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_rows =</span> F <span class="co"># specify row clustering</span></span>
<span id="cb453-4"><a href="heatmaps.html#cb453-4" aria-hidden="true" tabindex="-1"></a>         )</span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-93-1.png" width="672" /></p>
<p>Most heatmap packages have a clustering method argument in the function call.</p>
<div class="sourceCode" id="cb454"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb454-1"><a href="heatmaps.html#cb454-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, <span class="at">show_rownames =</span> F, <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>, <span class="at">col =</span>  col_palette, </span>
<span id="cb454-2"><a href="heatmaps.html#cb454-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_method =</span> <span class="st">&quot;complete&quot;</span>) <span class="co"># specify clustering method - select one of: ward.D&quot;, &quot;ward.D2&quot;, &quot;single&quot;, &quot;complete&quot;, &quot;average&quot; (= UPGMA), &quot;mcquitty&quot; (= WPGMA), &quot;median&quot; (= WPGMC) or &quot;centroid&quot; (= UPGMC)</span></span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-94-1.png" width="672" /></p>
<div class="sourceCode" id="cb455"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb455-1"><a href="heatmaps.html#cb455-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note: you can hide dendograms by setting arguments treeheight_row and/or treeheight_col to 0</span></span></code></pre></div>
<p>You can also make hierarchical cluster (hc) objects (i.e dendogram-like structures)
1. First make a distance matrix object using dist()</p>
<div class="sourceCode" id="cb456"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb456-1"><a href="heatmaps.html#cb456-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make distance matrices for columns (sample)</span></span>
<span id="cb456-2"><a href="heatmaps.html#cb456-2" aria-hidden="true" tabindex="-1"></a>dist_cols <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(exp_mat), <span class="at">method =</span> <span class="st">&quot;manhattan&quot;</span>) <span class="co">#default - select one of: one of &quot;euclidean&quot;, &quot;maximum&quot;, &quot;manhattan&quot;, &quot;canberra&quot;, &quot;binary&quot; or &quot;minkowski&quot;</span></span>
<span id="cb456-3"><a href="heatmaps.html#cb456-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb456-4"><a href="heatmaps.html#cb456-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Make distance matrices for rows (genes)</span></span>
<span id="cb456-5"><a href="heatmaps.html#cb456-5" aria-hidden="true" tabindex="-1"></a>dist_rows <span class="ot">&lt;-</span> <span class="fu">dist</span>(exp_mat, <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>) </span>
<span id="cb456-6"><a href="heatmaps.html#cb456-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb456-7"><a href="heatmaps.html#cb456-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Note: Can also make a correlation matrix into a distance matrix by subtracting values from 1</span></span>
<span id="cb456-8"><a href="heatmaps.html#cb456-8" aria-hidden="true" tabindex="-1"></a>dist_cols <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(<span class="dv">1</span> <span class="sc">-</span> <span class="fu">cor</span>(exp_mat, <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>))</span>
<span id="cb456-9"><a href="heatmaps.html#cb456-9" aria-hidden="true" tabindex="-1"></a>dist_rows <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(<span class="dv">1</span> <span class="sc">-</span> <span class="fu">cor</span>(<span class="fu">t</span>(exp_mat), <span class="at">method =</span> <span class="st">&quot;pearson&quot;</span>))</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Then build hierachal clustering objects using hclust()</li>
</ol>
<div class="sourceCode" id="cb457"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb457-1"><a href="heatmaps.html#cb457-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make hclust object for columns</span></span>
<span id="cb457-2"><a href="heatmaps.html#cb457-2" aria-hidden="true" tabindex="-1"></a>hc_cols <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dist_cols, <span class="at">method =</span> <span class="st">&quot;complete&quot;</span>)</span>
<span id="cb457-3"><a href="heatmaps.html#cb457-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc_cols)</span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-96-1.png" width="672" /></p>
<div class="sourceCode" id="cb458"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb458-1"><a href="heatmaps.html#cb458-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Repeat for rows</span></span>
<span id="cb458-2"><a href="heatmaps.html#cb458-2" aria-hidden="true" tabindex="-1"></a>hc_rows <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dist_rows, <span class="at">method =</span> <span class="st">&quot;complete&quot;</span>)</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Pass these objects into your heatmap function call</li>
</ol>
<div class="sourceCode" id="cb459"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb459-1"><a href="heatmaps.html#cb459-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, <span class="at">show_rownames =</span> F, <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>, <span class="at">col =</span>  col_palette, </span>
<span id="cb459-2"><a href="heatmaps.html#cb459-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols =</span> hc_cols, <span class="co"># specify column clustering</span></span>
<span id="cb459-3"><a href="heatmaps.html#cb459-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_rows =</span> hc_rows <span class="co"># specify row clustering</span></span>
<span id="cb459-4"><a href="heatmaps.html#cb459-4" aria-hidden="true" tabindex="-1"></a>         )</span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-97-1.png" width="672" /></p>
</div>
<div id="create-annotations" class="section level2" number="9.6">
<h2><span class="header-section-number">9.6</span> Create annotations</h2>
<ul>
<li><p>Annotations are important components of a heatmap that it shows additional information that associates with rows or columns in the heatmap.</p></li>
<li><p>Annotations in pheatmap are in the form of data frames.</p></li>
<li><p>Annotation column (sample information)</p></li>
</ul>
<div class="sourceCode" id="cb460"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb460-1"><a href="heatmaps.html#cb460-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get column names</span></span>
<span id="cb460-2"><a href="heatmaps.html#cb460-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(exp_mat)</span></code></pre></div>
<pre><code>##  [1] &quot;BC_E_1&quot;  &quot;BC_E_2&quot;  &quot;BC_EP_1&quot; &quot;BC_EP_2&quot; &quot;LP_E_1&quot;  &quot;LP_E_2&quot;  &quot;LP_EP_1&quot;
##  [8] &quot;LP_EP_2&quot; &quot;LM_E_1&quot;  &quot;LM_E_2&quot;  &quot;LM_EP_1&quot; &quot;LM_EP_2&quot;</code></pre>
<div class="sourceCode" id="cb462"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb462-1"><a href="heatmaps.html#cb462-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make annotation column</span></span>
<span id="cb462-2"><a href="heatmaps.html#cb462-2" aria-hidden="true" tabindex="-1"></a>ann_col <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">CellType =</span> <span class="fu">c</span>(<span class="st">&quot;BC&quot;</span>, <span class="st">&quot;BC&quot;</span>, <span class="st">&quot;BC&quot;</span>, <span class="st">&quot;BC&quot;</span>, <span class="st">&quot;LP&quot;</span>, <span class="st">&quot;LP&quot;</span>, <span class="st">&quot;LP&quot;</span>, <span class="st">&quot;LP&quot;</span>, <span class="st">&quot;LM&quot;</span>, <span class="st">&quot;LM&quot;</span>, <span class="st">&quot;LM&quot;</span>, <span class="st">&quot;LM&quot;</span>), </span>
<span id="cb462-3"><a href="heatmaps.html#cb462-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">Hormone =</span> <span class="fu">c</span>(<span class="st">&quot;E&quot;</span>, <span class="st">&quot;E&quot;</span>,  <span class="st">&quot;EP&quot;</span>, <span class="st">&quot;EP&quot;</span>, <span class="st">&quot;E&quot;</span>,  <span class="st">&quot;E&quot;</span>,  <span class="st">&quot;EP&quot;</span>, <span class="st">&quot;EP&quot;</span>, <span class="st">&quot;E&quot;</span>,  <span class="st">&quot;E&quot;</span>,  <span class="st">&quot;EP&quot;</span>, <span class="st">&quot;EP&quot;</span>))</span>
<span id="cb462-4"><a href="heatmaps.html#cb462-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename row names of annotation column dataframe as the column names of matrix (must match)</span></span>
<span id="cb462-5"><a href="heatmaps.html#cb462-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ann_col) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(exp_mat)</span></code></pre></div>
<p>Note: instead of passing in that huge vector for cell type, these lines of code do exactly the same thing</p>
<div class="sourceCode" id="cb463"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb463-1"><a href="heatmaps.html#cb463-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for cell type</span></span>
<span id="cb463-2"><a href="heatmaps.html#cb463-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;BC&quot;</span>, <span class="st">&quot;LP&quot;</span>, <span class="st">&quot;LM&quot;</span>), <span class="at">each =</span> <span class="dv">4</span>)</span></code></pre></div>
<pre><code>##  [1] &quot;BC&quot; &quot;BC&quot; &quot;BC&quot; &quot;BC&quot; &quot;LP&quot; &quot;LP&quot; &quot;LP&quot; &quot;LP&quot; &quot;LM&quot; &quot;LM&quot; &quot;LM&quot; &quot;LM&quot;</code></pre>
<div class="sourceCode" id="cb465"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb465-1"><a href="heatmaps.html#cb465-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gsub</span>(<span class="st">&quot;_.*&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="fu">colnames</span>(exp_mat))</span></code></pre></div>
<pre><code>##  [1] &quot;BC&quot; &quot;BC&quot; &quot;BC&quot; &quot;BC&quot; &quot;LP&quot; &quot;LP&quot; &quot;LP&quot; &quot;LP&quot; &quot;LM&quot; &quot;LM&quot; &quot;LM&quot; &quot;LM&quot;</code></pre>
<div class="sourceCode" id="cb467"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb467-1"><a href="heatmaps.html#cb467-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for hormone treatement</span></span>
<span id="cb467-2"><a href="heatmaps.html#cb467-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">&quot;E&quot;</span>, <span class="st">&quot;EP&quot;</span>), <span class="at">each =</span> <span class="dv">2</span>, <span class="at">times =</span> <span class="dv">3</span>)</span></code></pre></div>
<pre><code>##  [1] &quot;E&quot;  &quot;E&quot;  &quot;EP&quot; &quot;EP&quot; &quot;E&quot;  &quot;E&quot;  &quot;EP&quot; &quot;EP&quot; &quot;E&quot;  &quot;E&quot;  &quot;EP&quot; &quot;EP&quot;</code></pre>
<ul>
<li>Annotation rows (gene information)
If you have to provide annotation for features, make a data frame (just like annotation column) and rename rows as genes (same order as your matrix)</li>
</ul>
<div class="sourceCode" id="cb469"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb469-1"><a href="heatmaps.html#cb469-1" aria-hidden="true" tabindex="-1"></a>ann_row <span class="ot">&lt;-</span> <span class="cn">NA</span></span></code></pre></div>
<ul>
<li>Plot heatmap</li>
</ul>
<div class="sourceCode" id="cb470"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb470-1"><a href="heatmaps.html#cb470-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>, <span class="at">show_rownames =</span> F, <span class="at">col =</span> col_palette,</span>
<span id="cb470-2"><a href="heatmaps.html#cb470-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> ann_col, <span class="co"># annotation column (sample info)</span></span>
<span id="cb470-3"><a href="heatmaps.html#cb470-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> ann_row <span class="co"># annotation rows (gene info)</span></span>
<span id="cb470-4"><a href="heatmaps.html#cb470-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-101-1.png" width="672" /></p>
<p>Make colours for your annotations
- Make a list object of the annotation colors
- List elements have the same names as the columns of the annotation data frames
- Vector elements have same names as values</p>
<div class="sourceCode" id="cb471"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb471-1"><a href="heatmaps.html#cb471-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make annotation colours list</span></span>
<span id="cb471-2"><a href="heatmaps.html#cb471-2" aria-hidden="true" tabindex="-1"></a>ann_colors <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb471-3"><a href="heatmaps.html#cb471-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">CellType =</span>  <span class="fu">c</span>(<span class="at">BC=</span><span class="st">&quot;red&quot;</span>, <span class="at">LP=</span><span class="st">&quot;blue&quot;</span>, <span class="at">LM=</span><span class="st">&quot;darkgreen&quot;</span>),</span>
<span id="cb471-4"><a href="heatmaps.html#cb471-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">Hormone =</span> <span class="fu">c</span>(<span class="at">E =</span> <span class="st">&quot;black&quot;</span>, <span class="at">EP =</span> <span class="st">&quot;purple&quot;</span>)</span>
<span id="cb471-5"><a href="heatmaps.html#cb471-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb471-6"><a href="heatmaps.html#cb471-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb471-7"><a href="heatmaps.html#cb471-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot heatmap</span></span>
<span id="cb471-8"><a href="heatmaps.html#cb471-8" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>, <span class="at">show_rownames =</span> F, <span class="at">col =</span> col_palette,</span>
<span id="cb471-9"><a href="heatmaps.html#cb471-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_col =</span> ann_col, <span class="co"># annotation column (sample info)</span></span>
<span id="cb471-10"><a href="heatmaps.html#cb471-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> ann_row, <span class="co"># annotation rows (gene info)</span></span>
<span id="cb471-11"><a href="heatmaps.html#cb471-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_colors =</span> ann_colors <span class="co"># annotation colours</span></span>
<span id="cb471-12"><a href="heatmaps.html#cb471-12" aria-hidden="true" tabindex="-1"></a>         )</span></code></pre></div>
<p><img src="r-sessions_files/figure-html/unnamed-chunk-102-1.png" width="672" /></p>
<p>Save pheatmap to file
- Add title using main =
- Save to file using filename = (Currently following formats are supported: png, pdf, tiff, bmp, jpeg)</p>
<div class="sourceCode" id="cb472"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb472-1"><a href="heatmaps.html#cb472-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot heatmap and save to file</span></span>
<span id="cb472-2"><a href="heatmaps.html#cb472-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(exp_mat, <span class="at">scale =</span> <span class="st">&quot;row&quot;</span>, <span class="at">show_rownames =</span> F, <span class="at">col =</span> col_palette, <span class="at">annotation_col =</span> ann_col, <span class="at">annotation_colors =</span> ann_colors,</span>
<span id="cb472-3"><a href="heatmaps.html#cb472-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">&quot;My heatmap&quot;</span>,<span class="co"># add a title</span></span>
<span id="cb472-4"><a href="heatmaps.html#cb472-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">filename =</span> <span class="st">&quot;my_heatmap.png&quot;</span>) <span class="co"># save as png to current working directory</span></span></code></pre></div>
</div>
<div id="practice-4" class="section level2" number="9.7">
<h2><span class="header-section-number">9.7</span> Practice</h2>
<p>Motor trend car road tests (mtcars)
Fuel consumption and 10 additional aspects (variables) of automobile design tested in 1974 are given in an in-built R dataset called “mtcars”.
a) Look at the first rows of the data frame (using head())
b) Look at the structure, are all the data numeric? (use str())
c) Convert the data frame to a data matrix into a new variable.
d) Add 1 to all values and apply log2 transformation.
e) Make a heatmap (using the pheatmap package).
i) Scale by column,
ii) use clustering method “ward.D”,
iii) set cell width to 10,
iv) pick 9 colours from RColorBrewer color palette “RdPu” and make a palette with 250 colors for the heatmap, and
v) set the title to “my heatmap”
vi) save to a jpeg file called “mtcars_heatmap”</p>
<p>Solution</p>
<div class="sourceCode" id="cb473"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb473-1"><a href="heatmaps.html#cb473-1" aria-hidden="true" tabindex="-1"></a><span class="co"># a) Use head() </span></span>
<span id="cb473-2"><a href="heatmaps.html#cb473-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mtcars)</span></code></pre></div>
<pre><code>##                    mpg cyl disp  hp drat    wt  qsec vs am gear carb
## Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
## Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
## Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
## Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
## Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
## Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1</code></pre>
<div class="sourceCode" id="cb475"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb475-1"><a href="heatmaps.html#cb475-1" aria-hidden="true" tabindex="-1"></a><span class="co"># b) Use str() to see the structure</span></span>
<span id="cb475-2"><a href="heatmaps.html#cb475-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(mtcars) <span class="co"># ANSWER: Yes, all variables are numeric</span></span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    32 obs. of  11 variables:
##  $ mpg : num  21 21 22.8 21.4 18.7 18.1 14.3 24.4 22.8 19.2 ...
##  $ cyl : num  6 6 4 6 8 6 8 4 4 6 ...
##  $ disp: num  160 160 108 258 360 ...
##  $ hp  : num  110 110 93 110 175 105 245 62 95 123 ...
##  $ drat: num  3.9 3.9 3.85 3.08 3.15 2.76 3.21 3.69 3.92 3.92 ...
##  $ wt  : num  2.62 2.88 2.32 3.21 3.44 ...
##  $ qsec: num  16.5 17 18.6 19.4 17 ...
##  $ vs  : num  0 0 1 1 0 1 0 1 1 1 ...
##  $ am  : num  1 1 1 0 0 0 0 0 0 0 ...
##  $ gear: num  4 4 4 3 3 3 3 4 4 4 ...
##  $ carb: num  4 4 1 1 2 1 4 2 2 4 ...</code></pre>
<div class="sourceCode" id="cb477"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb477-1"><a href="heatmaps.html#cb477-1" aria-hidden="true" tabindex="-1"></a><span class="co"># c) convert to matrix using as.matrix</span></span>
<span id="cb477-2"><a href="heatmaps.html#cb477-2" aria-hidden="true" tabindex="-1"></a>mtcars_mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(mtcars)</span>
<span id="cb477-3"><a href="heatmaps.html#cb477-3" aria-hidden="true" tabindex="-1"></a><span class="co"># d) add 1 and log2</span></span>
<span id="cb477-4"><a href="heatmaps.html#cb477-4" aria-hidden="true" tabindex="-1"></a>mtcars_mat <span class="ot">&lt;-</span> mtcars_mat <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb477-5"><a href="heatmaps.html#cb477-5" aria-hidden="true" tabindex="-1"></a>mtcars_mat <span class="ot">&lt;-</span> <span class="fu">log2</span>(mtcars_mat)</span>
<span id="cb477-6"><a href="heatmaps.html#cb477-6" aria-hidden="true" tabindex="-1"></a><span class="co"># equivalent: mtcars_mat &lt;- log2(mtcars_mat + 1)</span></span>
<span id="cb477-7"><a href="heatmaps.html#cb477-7" aria-hidden="true" tabindex="-1"></a><span class="co"># e) plot heatmap</span></span>
<span id="cb477-8"><a href="heatmaps.html#cb477-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb477-9"><a href="heatmaps.html#cb477-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb477-10"><a href="heatmaps.html#cb477-10" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(mtcars_mat, <span class="co">#e</span></span>
<span id="cb477-11"><a href="heatmaps.html#cb477-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">scale =</span> <span class="st">&quot;column&quot;</span>, <span class="co">#i</span></span>
<span id="cb477-12"><a href="heatmaps.html#cb477-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">clustering_method =</span> <span class="st">&quot;ward.D&quot;</span>, <span class="co">#ii</span></span>
<span id="cb477-13"><a href="heatmaps.html#cb477-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">cellwidth =</span> <span class="dv">10</span>, <span class="co">#iii</span></span>
<span id="cb477-14"><a href="heatmaps.html#cb477-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">col =</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;RdPu&quot;</span>)) (<span class="dv">256</span>), <span class="co">#iv</span></span>
<span id="cb477-15"><a href="heatmaps.html#cb477-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">main =</span> <span class="st">&quot;my_heatmap&quot;</span>, <span class="co">#v</span></span>
<span id="cb477-16"><a href="heatmaps.html#cb477-16" aria-hidden="true" tabindex="-1"></a>         <span class="at">filename =</span> <span class="st">&quot;mtcars_heatmap.jpeg&quot;</span> <span class="co">#vi</span></span>
<span id="cb477-17"><a href="heatmaps.html#cb477-17" aria-hidden="true" tabindex="-1"></a>         )</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="correlation-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="dataframe-manipulation.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
